blockchain/chain.go

```diff
@@ -618,21 +617,21 @@ func (b *BlockChain) pruneStakeNodes() {
 		return
 	}
 
-	// Push the nodes to delete on a list in reverse order since it's easier
-	// to prune them going forwards than it is backwards.  This will
-	// typically end up being a single node since pruning is currently done
-	// just before each new node is created.  However, that might be tuned
-	// later to only prune at intervals, so the code needs to account for
-	// the possibility of multiple nodes.
-	deleteNodes := list.New()
+	// Push the nodes to delete on a list. This will typically end up being
+	// a single node since pruning is currently done just before each new
+	// node is created.  However, that might be tuned later to only prune at
+	// intervals, so the code needs to account for the possibility of
+	// multiple nodes.
+	var deleteNodes []*blockNode
```

Comment:
Given that we're always pruning from `pruneToNode` all the way back to genesis, could this be initialized with `deleteNodes := make([]*blockNode, 0, pruneToNode.height)?

Though I should say, this whole pruning process seems to be doing more work than it potentially could get away with and doesn't seem to prune from sidechains (unless that is done somewhere else that I haven't looked).

Reply:
As the comment indicates, this will almost always just be a  single node, so it seems like allocating space for whatever the current height of the chain's worth of space would be quite wasteful.

I agree that the pruning in general here needs to be revisited as it's far less efficient than it could be.